src/main.rs

```diff
@@ -1,29 +1,78 @@
-use std::hash::Hash;
-use std::io::ErrorKind;
-use std::iter::FromIterator;
-use std::{
-    collections::HashSet,
-    fs::{self, File},
-    process::Command,
-};
-
-use anyhow::{bail, Error, Result};
+mod handler;
+use anyhow::{bail, Result};
 use clap::{Parser, Subcommand, ValueEnum};
+use config::{Config, File, FileFormat};
 use glob::glob;
-use serde::{Deserialize, Serialize};
+use handler::*;
+use serde::Deserialize;
+use std::path::Path;
+use std::process::Command;
+use std::str;
+
+/// dir that greenboot looks for the health check and other scripts
+static GREENBOOT_INSTALL_PATHS: [&str; 2] = ["/usr/lib/greenboot", "/etc/greenboot"];
+
+/// greenboot config path
+static GREENBOOT_CONFIG_FILE: &str = "/etc/greenboot/greenboot.conf";
 
 #[derive(Parser)]
 #[clap(author, version, about, long_about = None)]
 #[clap(propagate_version = true)]
+
+/// cli parameters for greenboot
 struct Cli {
     #[clap(value_enum, short, long, default_value_t = LogLevel::Info)]
     log_level: LogLevel,
-
     #[clap(subcommand)]
     command: Commands,
 }
+#[derive(Debug, Deserialize)]
+///config params for greenboot
+struct GreenbootConfig {
+    max_reboot: i32,
```

Comment:
I've forgotten, why this needed to be an integer instead of an unsigned integer?

Reply:
I don't remember either,  `max_reboot` updates `boot_counter` where -1  executes the rollback.
usecae: if any failure rollback immediately.

Reply:
I don't think that would be a feature, if a user is configuring a maximum, that maximum should be always positive or 0.

Reply:
I dont fully agree as -1 is hard_rollback from osbuild level.

Reply:
I missed that, can I get a link to that?

Reply:
let me see if I can find it some where in documentation, AFAIK its the reason the boot_decrement counter  is  handled at low level to prevent infinite restart cycle even greenboot fails .

Reply:
still looking for the documentation, meanwhile
```
[core@localhost ~]$ sudo grub2-editenv - set boot_success=0
[core@localhost ~]$ sudo grub2-editenv list
boot_success=0
[core@localhost ~]$ sudo grub2-editenv - set boot_counter=-1
[core@localhost ~]$ sudo grub2-editenv list
boot_success=0
boot_counter=-1
[core@localhost ~]$ rpm-ostree status
State: idle
Deployments:
● rhel-edge:rhel/9/x86_64/edge
                  Version: 9.4 (2024-04-15T08:44:16Z)
                   Commit: 6ebbc4e8ea95bc80207fbe1a7e950780ab1bd1eac5853d728d4dbccf6c5addc2
                 Unlocked: hotfix

  rhel-edge:rhel/9/x86_64/edge
                  Version: 9.4 (2024-04-15T08:44:16Z)
                   Commit: 6ebbc4e8ea95bc80207fbe1a7e950780ab1bd1eac5853d728d4dbccf6c5addc2
[core@localhost ~]$ sudo reboot
[core@localhost ~]$ Connection to 192.168.124.128 closed by remote host.
Connection to 192.168.124.128 closed.
[~/git_repos/composer_assets]$
λ ssh core@192.168.124.128
Greenboot healthcheck passed - status is GREEN.
Last login: Thu Apr 18 04:25:13 2024 from 192.168.124.1
[core@localhost ~]$ rpm-ostree status
State: idle
Deployments:
  rhel-edge:rhel/9/x86_64/edge
                  Version: 9.4 (2024-04-15T08:44:16Z)
                   Commit: 6ebbc4e8ea95bc80207fbe1a7e950780ab1bd1eac5853d728d4dbccf6c5addc2
                 Unlocked: hotfix

● rhel-edge:rhel/9/x86_64/edge
                  Version: 9.4 (2024-04-15T08:44:16Z)
                   Commit: 6ebbc4e8ea95bc80207fbe1a7e950780ab1bd1eac5853d728d4dbccf6c5addc2
[core@localhost ~]$ 
```

Reply:
It doesn't rollback but switches back to the previous version on next reboot.

Reply:
so the difference is that if we set a -1 instead of rolling back and deleting the previous deployment it switches back to the previous one without deleting the buggy deployment?

Reply:
thats ostree way of stopping the floodgate if greenboot fails to execute.